<a href="https://colab.research.google.com/github/PasornC/MORS950/blob/main/Assignment1_MORS950_HMI.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import pandas as pd

url_with_data = "https://mattgroh.com/data/assignment1_customerdata.csv"
df = pd.read_csv(url_with_data)

# Ask ChatGPT or your favorite LLM to help you answer the questions.
# Treat the LLM like a smart intern who is highly capable at coding but prone to surprising errors.
# Ask lots of questions to the LLM when you don't know something, and ask for explanations.
# When things go wrong (e.g. Colab notebook telling you there is an error), ask the LLM, "Are you sure?"

In [3]:
print("This dataframe has {} rows and {} columns".format(df.shape[0], df.shape[1]))
print(df.columns)

This dataframe has 7043 rows and 23 columns
Index(['customer_id', 'gender', 'senior_citizen', 'partner', 'dependents',
       'tenure', 'phone_service', 'multiple_lines', 'internet_service',
       'online_security', 'online_backup', 'device_protection', 'tech_support',
       'streaming_tv', 'streaming_movies', 'contract', 'paperless_billing',
       'payment_method', 'monthly_charges', 'total_charges', 'churn',
       'churn_prediction_ml_model_a', 'churn_prediction_ml_model_b'],
      dtype='object')


In [ ]:
df.head()

customer_id  gender  senior_citizen  partner  dependents  tenure  \
0            1       0               0        1           0       1   
1            2       1               0        0           0      34   
2            3       1               0        0           0       2   
3            4       1               0        0           0      45   
4            5       0               0        0           0       2   

   phone_service  multiple_lines  internet_service  online_security  ...  \
0              0               1                 0                0  ...   
1              1               0                 0                2  ...   
2              1               0                 0                2  ...   
3              0               1                 0                2  ...   
4              1               0                 1                0  ...   

   streaming_tv  streaming_movies  contract  paperless_billing  \
0             0                 0         0                  1   
1             0                 0         1                  0   
2             0                 0         0                  1   
3             0                 0         1                  0   
4             0                 0         0                  1   

   payment_method  monthly_charges  total_charges  churn  \
0               2            29.85           2505      0   
1               3            56.95           1466      0   
2               3            53.85            157      1   
3               0            42.30           1400      0   
4               2            70.70            925      1   

   churn_prediction_ml_model_a  churn_prediction_ml_model_b  
0                     0.912821                     0.415385  
1                     0.538462                     0.102564  
2                     0.428205                     0.325641  
3                     0.705128                     0.537180  
4                     0.743590                     0.229487  

[5 rows x 23 columns]

Q1. Before looking at the data, what features do you think will be most important for predicting churn, and why?

I think "customer_id" will be the most important because this is the unique feature that can allow me to calculate the churn rate by dividing the total number of customer at the end of the period with total number of the customer at the begining of the period.

Q2. What percent of customers churn?

In [5]:
import pandas as pd

# Load the data
url_with_data = "https://mattgroh.com/data/assignment1_customerdata.csv"
df = pd.read_csv(url_with_data)

# Assuming 'churn' is the column representing whether a customer has churned
churned_customers = df[df['churn'] == 1].shape[0]
total_customers = df.shape[0]

# Calculate the churn rate
churn_rate = churned_customers / total_customers

# Print the churn rate
print(f"The churn rate is: {churn_rate * 100:.2f}%")

The churn rate is: 26.54%


Q3. How accurately could you predict customer churn with random guessing?

If the industry has a churn rate that is consistent then it might be possible to predict the churn rate with 60%-70% accuracy, or else it is very unlikely to predict the churn rate accurately.

Q4. What percent of senior citizens churn?

In [6]:
import pandas as pd

# Load the data
url_with_data = "https://mattgroh.com/data/assignment1_customerdata.csv"
df = pd.read_csv(url_with_data)

# Assuming 'churn' is the column representing whether a customer has churned
# Assuming 'senior_citizen' is the column representing whether a customer is a senior citizen
senior_churned_customers = df[(df['churn'] == 1) & (df['senior_citizen'] == 1)].shape[0]
total_senior_customers = df[df['senior_citizen'] == 1].shape[0]

# Calculate the churn rate for senior citizens
senior_churn_rate = senior_churned_customers / total_senior_customers

# Print the churn rate for senior citizens
print(f"The churn rate for senior citizens is: {senior_churn_rate * 100:.2f}%")

The churn rate for senior citizens is: 41.68%


Q5. Split the data into a training set with 80% of the data and a test set with 20% of the data.
How accurately can you predict churn with a linear regression? What about with a
random forest?

In [7]:
import pandas as pd
from sklearn.model_selection import train_test_split

# Load the data
url_with_data = "https://mattgroh.com/data/assignment1_customerdata.csv"
df = pd.read_csv(url_with_data)

# Assuming 'churn' is the column representing whether a customer has churned
# Assuming other_columns is a list of other feature columns in your dataset

# Selecting features and target variable
numerical_columns = df.select_dtypes(include='number').columns
other_columns = numerical_columns.drop(['churn'])
X = df[other_columns]  # Features
y = df['churn']  # Target variable

# Split the data into a training set (80%) and a test set (20%)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Print the shapes of the resulting sets
print(f"Shape of X_train: {X_train.shape}")
print(f"Shape of X_test: {X_test.shape}")
print(f"Shape of y_train: {y_train.shape}")
print(f"Shape of y_test: {y_test.shape}")

Shape of X_train: (5634, 22)
Shape of X_test: (1409, 22)
Shape of y_train: (5634,)
Shape of y_test: (1409,)


In [9]:
#Linear Regression
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

# Initialize the linear regression model
linear_model = LinearRegression()

# Train the model on the training set
linear_model.fit(X_train, y_train)

# Predict churn on the test set
linear_predictions = linear_model.predict(X_test)

# Calculate and print the mean squared error
linear_mse = mean_squared_error(y_test, linear_predictions)
print(f"Linear Regression Mean Squared Error: {linear_mse}")

# Calculate and print the predicted churn rate
linear_churn_rate = linear_predictions.mean()
print(f"Linear Regression Predicted Churn Rate: {linear_churn_rate * 100:.2f}%")

Linear Regression Mean Squared Error: 0.1378752313221354
Linear Regression Predicted Churn Rate: 26.27%


In [10]:
#Random Forest
from sklearn.ensemble import RandomForestRegressor

# Initialize the random forest model
rf_model = RandomForestRegressor(random_state=42)

# Train the model on the training set
rf_model.fit(X_train, y_train)

# Predict churn on the test set
rf_predictions = rf_model.predict(X_test)

# Calculate and print the mean squared error
rf_mse = mean_squared_error(y_test, rf_predictions)
print(f"Random Forest Mean Squared Error: {rf_mse}")

# Calculate and print the predicted churn rate
rf_churn_rate = rf_predictions.mean()
print(f"Random Forest Predicted Churn Rate: {rf_churn_rate * 100:.2f}%")

Random Forest Mean Squared Error: 0.1416845990063875
Random Forest Predicted Churn Rate: 27.32%


Both methods provide highly accurate result, within ±1% of the correct churn rate - 26.54%

Q6. What features did you use to predict churn?

I use 'gender'
'senior_citizen'
'partner'
'dependents'
'tenure'
'phone_service'
'multiple_lines'
'internet_service'
'online_security'
'online_backup'
'device_protection'
'tech_support'
'streaming_tv'
'streaming_movies'
'contract'
'paperless_billing'
'payment_method'
'monthly_charges'
'total_charges'

Q7. What are the most important features for predicting churn with a linear regression (share
the top three in order of importance)? What about with a random forest (share the top
three in order of importance)?

In [13]:
#Linear Regression
from sklearn.linear_model import LinearRegression
import pandas as pd

# Assuming 'df' is your original DataFrame
df_encoded = pd.get_dummies(df, columns=['gender', 'partner', 'dependents', 'multiple_lines', 'internet_service', 'online_security', 'online_backup', 'device_protection', 'tech_support', 'streaming_tv', 'streaming_movies', 'contract', 'paperless_billing', 'payment_method'], drop_first=True)

# Assuming 'df_encoded' is the DataFrame with one-hot encoded categorical variables
X_linear = df_encoded.drop(['customer_id', 'churn'], axis=1)  # Features
y_linear = df_encoded['churn']  # Target variable

# Initialize the linear regression model
linear_model = LinearRegression()

# Train the model on the entire dataset
linear_model.fit(X_linear, y_linear)

# Get feature coefficients
linear_coefficients = linear_model.coef_

# Pair features with coefficients
feature_coefficients_linear = list(zip(X_linear.columns, linear_coefficients))

# Sort features by absolute coefficient value in descending order
sorted_feature_coefficients_linear = sorted(feature_coefficients_linear, key=lambda x: abs(x[1]), reverse=True)

# Print the top three features for Linear Regression
print("Top three features for Linear Regression:")
for feature, coefficient in sorted_feature_coefficients_linear[:3]:
    print(f"{feature}: {coefficient}")

Top three features for Linear Regression:
internet_service_1: 0.20799661763614347
contract_1: -0.10696962096181677
payment_method_2: 0.07053740110483078


In [14]:
#Random Forest
from sklearn.ensemble import RandomForestRegressor
import pandas as pd

# Assuming 'df' is the original DataFrame with categorical variables
df_encoded_rf = pd.get_dummies(df, columns=['gender', 'partner', 'dependents', 'multiple_lines', 'internet_service', 'online_security', 'online_backup', 'device_protection', 'tech_support', 'streaming_tv', 'streaming_movies', 'contract', 'paperless_billing', 'payment_method'], drop_first=True)

# Assuming 'df_encoded_rf' is the DataFrame with one-hot encoded categorical variables
X_rf = df_encoded_rf.drop(['customer_id', 'churn'], axis=1)  # Features
y_rf = df_encoded_rf['churn']  # Target variable

# Initialize the random forest model
rf_model = RandomForestRegressor(random_state=42)

# Train the model on the entire dataset
rf_model.fit(X_rf, y_rf)

# Get feature importances
rf_importances = rf_model.feature_importances_

# Pair features with importances
feature_importances_rf = list(zip(X_rf.columns, rf_importances))

# Sort features by importance in descending order
sorted_feature_importances_rf = sorted(feature_importances_rf, key=lambda x: x[1], reverse=True)

# Print the top three features for Random Forest
print("\nTop three features for Random Forest:")
for feature, importance in sorted_feature_importances_rf[:3]:
    print(f"{feature}: {importance}")


Top three features for Random Forest:
tenure: 0.19878364589530456
total_charges: 0.12452396728663478
monthly_charges: 0.12347048641612862


Q8. How often did your predictions beat "Model A"? What about "Model B"? These
predictions are found in the last two columns of the .csv.